### PR TITLE
IRON-76

### PR DESCRIPTION
**Pull request by gmaf**
Add mobile authorization, based on openid connect

### CODE REVIEW

File: src/Lykke.Service.IroncladDecorator/Controllers/CallbackController.cs [hunk 1]
```diff
@@ -112,13 +112,17 @@ public IActionResult AfterLoginIos()
             return Ok();
         }
 
-        private async Task ProcessMobileCallback(IdentityProviderClientSettings clientSettings, string redirectUri)
+        private async Task ProcessMobileCallback()
         {
-            var userSession = new UserSession.UserSession();
-
-            var authCode = HttpContext.Request.Query["code"];
+            if (!HttpContext.Items.TryGetValue("CallbackTokenEndpointResponse", out var tokensValue))
```
Comment (mskobzev-old) [inline src/Lykke.Service.IroncladDecorator/Controllers/CallbackController.cs]: Please extract string const

File: src/Lykke.Service.IroncladDecorator/Controllers/CallbackController.cs [hunk 2]
```diff
@@ -112,13 +112,17 @@ public IActionResult AfterLoginIos()
...
+            {
+                throw new Exception("No token result in callback.");
+            }
 
-            var tokens = await GetTokens(authCode, clientSettings, redirectUri);
+            if(!(tokensValue is TokenData tokens))
+                throw new Exception("Could not cast to token data.");
+            
+            var userSession = await _userSessionManager.GetUserSession() ?? new UserSession.UserSession();
```
Comment (mskobzev-old) [inline src/Lykke.Service.IroncladDecorator/Controllers/CallbackController.cs]: UserSession.UserSession looks wierd.. maybe you could just rename namespace?

File: src/Lykke.Service.IroncladDecorator/OpenIdConnect/CustomOpenIdConnectEvents.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+            var properties = context.Properties;
+            
+            if (!string.IsNullOrEmpty(protocolMessage.State))
+                properties.Items[OpenIdConnectDefaults.UserstatePropertiesKey] = protocolMessage.State;
+            properties.Items[OpenIdConnectDefaults.RedirectUriForCodePropertiesKey] = protocolMessage.RedirectUri;
+            protocolMessage.State = context.Options.StateDataFormat.Protect(properties);
+
+            //note: save correct auth url to redirect to it in ConnectController.
+            context.HttpContext.Items.Add("OpenIdConnectAuthenticationRequestUrl", protocolMessage.CreateAuthenticationRequestUrl());
```
Comment (mskobzev-old) [inline src/Lykke.Service.IroncladDecorator/OpenIdConnect/CustomOpenIdConnectEvents.cs]: Please extract string const

**Review by mskobzev-old — APPROVED** [review]

File: src/Lykke.Service.IroncladDecorator/OpenIdConnect/CustomOpenIdConnectEvents.cs (hunk 3 above)
Comment (gmaf) [reply src/Lykke.Service.IroncladDecorator/OpenIdConnect/CustomOpenIdConnectEvents.cs]: Fixed

**Pull request merged**